lib/trace/store.js

```diff
@@ -0,0 +1,143 @@
+    this.tracers = {};
+}
+require('util').inherits(TracerStore, EventEmitter);
+
+TracerStore.prototype.add = function add(config, opts, callback) {
+    var self = this;
+    
+    if (callback) {
+        self.once('add', callback);
+    }
+
+    var err = core.Tracer.invalidateOptions(opts);
+    if (err) {
+        self.emit('add', err);
+        return null;
+    }
+
+    var tracer = self._get(config, opts);
+    if (!tracer) {
+        var TracerType = types[opts.sink.type];
+        tracer = new TracerType(self.ringpop, config, opts);
+        var err = tracer.invalidate(opts);
+        if (err) {
+            self.emit('add', err);
+            return null;
+        }
+
+        var eventTracers = self.tracers[config.traceEvent];
+        if (!eventTracers) {
+            eventTracers = [];
+            self.tracers[config.traceEvent] = eventTracers;
+        }
+        eventTracers.push(tracer);
+
+        tracer.connectSink(function onConnect(err) {
+            if (err) {
+                self.remove(config, opts, null);
+                self.emit('add', err, null);
+                return;
+            }
+            tracer.connectSource();
+            self._updateExpiration(config, opts, tracer);
+            self.emit('add', null, tracer);
+        });
+    } else {
+        self._updateExpiration(config, opts, tracer);
+        self.emit('add', null, tracer);
+    }
+
+    return tracer;
+}
+
+TracerStore.prototype.remove = function remove(config, opts, callback) {
+    if (callback) {
+        this.once('remove', callback);
+    }
+
+    var tracer = null;
+    var i = this._indexOf(config, opts);
+    if (i >= 0) {
+        tracer = this.tracers[config.traceEvent][i];
+    }
+    if (tracer) {
+        this.tracers[config.traceEvent].splice(i, 1);
+        tracer.disconnectSink();
+        tracer.disconnectSource();
+    }
+    this.emit('remove', null, tracer);
+    return tracer;
+}
+
+TracerStore.prototype._indexOf = function _indexOf(config, opts) {
+    var traceEvent = config.traceEvent;
+    var tracers = this.tracers[traceEvent] || [];
+    for (var i = 0; i < tracers.length; ++i) {
+        var tracer = tracers[i];
+        if (tracer.matches(traceEvent, opts)) {
+            return i;
+        }
+    }
+    return -1;
+}
+
+TracerStore.prototype._get = function _get(config, opts) {
+    var i = this._indexOf(config, opts);
+    return i >= 0 ? this.tracers[config.traceEvent][i] : null;
+}
+
+TracerStore.prototype._updateExpiration = function _updateExpiration(config, opts, tracer) {
+    var now = Date.now();
+    var offset = Math.max(0, opts.expiresIn || (opts.expiresAt - now));
+    var timer = tracer._timer;
+    var remove = this.remove.bind(this, config, opts);
+
+    // 0/null offset implies already expired, thus don't re-add
+    if (timer) {
+        clearTimeout(tracer._timer);
+        tracer._timer = null;
+    }
+    if (offset) {
+        tracer._timer = setTimeout(remove, offset);
```

Comment:
TracerStore should offer up a `destroy()` function to clear all existing tracer timers. TracerStore can listen to the `destroyed` event that Ringpop emits or call tracerStore.destroy() from Ringpop's destroy() function. The former is the convention I've been using lately.